### ISSUE TITLE
Negative and cross-group rewards appear to be canceled out by GRPO advantage normalization in supervised GRACE training

### ISSUE THREAD

**Issue by turnaround5954** [GasolSun36/GRACE#4]
Hi, thanks for releasing GRACE. I have been studying the supervised GRACE training implementation and I have a question about whether the negative and cross-group reward terms actually contribute to the GRPO policy gradient.

From my reading of the current code, it looks like `negative_sim_rewards` and likely `cross_group_rewards` become constant within each GRPO group, so they are canceled out when GRPO computes per-group relative advantages. As a result, only the reward components that vary across the `rollout.n` sampled positive/main responses, such as `positive_sim_rewards`, `clustering_rewards`, and possibly `length_penalty`, seem to affect the final policy update.

Could you confirm whether this is intended?

### Relevant training configuration

The default `train.sh` uses GRPO with multiple rollouts:

```bash
algorithm.adv_estimator=grpo
algorithm.norm_adv_by_std_in_grpo=False
actor_rollout_ref.rollout.n=8
reward_model.enable=False
reward_model.reward_manager=hidden
+data.train_mode=supervised
```

So the effective policy signal should come from reward differences among the `rollout.n` responses sharing the same `uid`.

### What I observed in the rollout path

In `verl/workers/rollout/vllm_rollout/vllm_rollout_spmd.py`, the main positive/document prompt is sampled with `self.sampling_params`, whose `n` is configured as `rollout.n`.

However, query and negative document generations are forced to `n = 1`:

```python
query_sampling_params = copy.deepcopy(self.sampling_params)
query_sampling_params.n = 1
query_sampling_params.temperature = 0.5

if has_negative:
    negative_doc_sampling_params = copy.deepcopy(self.sampling_params)
    negative_doc_sampling_params.n = 1
    negative_doc_sampling_params.temperature = 0.5
```

Later, when `self.sampling_params.n > 1`, the query and negative document sequences are repeated to match the expanded main responses:

```python
query_idx = _repeat_interleave(query_idx, self.sampling_params.n)
query_attention_mask = _repeat_interleave(query_attention_mask, self.sampling_params.n)
query_position_ids = _repeat_interleave(query_position_ids, self.sampling_params.n)
query_response = _repeat_interleave(query_response, self.sampling_params.n)

if has_negative:
    negative_doc_idx = _repeat_interleave(negative_doc_idx, self.sampling_params.n)
    negative_doc_attention_mask = _repeat_interleave(
        negative_doc_attention_mask, self.sampling_params.n
    )
    negative_doc_position_ids = _repeat_interleave(
        negative_doc_position_ids, self.sampling_params.n
    )
    negative_doc_response = _repeat_interleave(
        negative_doc_response, self.sampling_params.n
    )
```

Therefore, within one original training example / one GRPO group, the positive/main responses differ across the `n` samples, but the query response and negative document response are identical copies.

### What I observed in the reward function

In `verl/workers/reward_manager/grace.py`, rewards are grouped by the prompt tensor. For supervised training, this corresponds to the repeated main/positive prompt group.

The positive reward is computed per sample:

```python
group_positive_sim = self.compute_similarity(
    group_query, group_main, similarity_method
)
positive_rewards[start_idx:end_idx] = group_positive_sim
```

Since `group_main` differs across rollouts, this term can vary within the GRPO group.

But the negative reward appears to be computed only from the first query/negative pair in the group and then broadcast to every sample in that group:

```python
group_negative_sim = self.compute_similarity(
    group_query[0:1], group_neg[0:1], similarity_method
)[0]

negative_rewards[start_idx:end_idx] = -group_negative_sim
```

This means:

```text
negative_rewards[i] = constant_for_this_grpo_group
```

The final reward includes this term:

```python
total_rewards = positive_rewards + negative_rewards
```

or, with temperature scaling:

```python
total_rewards = positive_sim_rewards + negative_sim_rewards
```

So the negative term is present in the raw reward tensor, but it is constant across all candidates in the same GRPO group.

### Why I think GRPO cancels this term

In `verl/trainer/ppo/core_algos.py`, GRPO computes sequence-level scores and then normalizes them by group:

```python
scores = token_level_rewards.sum(dim=-1)

for i in range(bsz):
    id2score[index[i]].append(scores[i])

for i in range(bsz):
    if norm_adv_by_std_in_grpo:
        scores[i] = (scores[i] - id2mean[index[i]]) / (
            id2std[index[i]] + epsilon
        )
    else:
        scores[i] = scores[i] - id2mean[index[i]]

scores = scores.unsqueeze(-1) * response_mask
```

For one GRPO group, if the reward is:

```text
R_i = positive_i + negative_const
```

then GRPO computes:

```text
A_i = R_i - mean(R)
    = (positive_i + negative_const)
      - mean(positive_i + negative_const)
    = positive_i - mean(positive_i)
```

So the negative constant cancels exactly. With std normalization enabled, adding a constant also does not change the normalized advantages.

### Cross-group reward may have the same issue

The cross-group term also seems likely to be constant within the same GRPO group.

In `_compute_rewards_on_reordered_data`, cross-group reward is computed from `current_query` against main representations from other groups:

```python
current_query = group_query[i : i + 1]

group_sims = self.compute_similarity(
    current_query.expand(other_group_mains.shape[0], -1),
    other_group_mains,
    similarity_method,
)

max_sim = torch.max(group_sims)
...
cross_group_reward_avg = -torch.stack(max_sims).mean()
cross_group_rewards[start_idx + i] = cross_group_reward_avg
```

But because `query_response` is generated once and then repeated across the `rollout.n` samples, `group_query[i]` is identical within the same GRPO group. The set of `other_group_mains` is also the same for all samples in that group. Therefore `cross_group_rewards` also appears to be constant within the group and should be canceled by GRPO advantage normalization.

### Policy update only optimizes the main response

Another related observation: the actor update only selects the main response fields:

```python
select_keys = [
    "responses",
    "input_ids",
    "attention_mask",
    "position_ids",
    "old_log_probs",
    "advantages",
]
```

So even though query and negative document responses are generated and their hidden states are used for reward computation, their token log-probabilities are not directly optimized by the PPO/GRPO policy loss. The policy gradient is applied only to the main `responses`.

### Current effective reward signal, as I understand it

Under the default supervised GRACE setup, the reward components that can affect GRPO advantage are the ones that vary among the `rollout.n` main responses for the same `uid`.

That means the effective policy signal seems to come mainly from:

```text
positive_sim_rewards
+ clustering_weight * clustering_rewards
+ length_penalty, if triggered differently across candidates
```

while these terms are included in raw reward logging but seem canceled in GRPO advantage:

```text
negative_sim_rewards
cross_group_rewards
```

### Question

Is this behavior intended?

If the goal is for the negative document to actively affect the GRPO policy update, should the implementation instead make the negative-related reward vary across the `rollout.n` candidates, for example by computing negative similarity using each candidate’s main response against the negative document, or by sampling/optimizing negative responses per rollout rather than generating once and repeating?

Thanks again for the release. I wanted to check whether my reading is correct before assuming this is a bug.
